### DISCUSSION TITLE
CodeGen package confusion: As of *02-2023*, what officially to install to use them in Net7?

### DISCUSSION THREAD

**Discussion by YkTru** [louthy/language-ext#1195] [General]
I followed actual documentation as best as I could:

I installed: `NuGet\Install-Package CodeGeneration.Roslyn.BuildTime -Version 0.6.1`

Project file:
`<PackageReference Include="LanguageExt.CodeGen" Version="4.4.2" />`
`<PackageReference Include="CodeGeneration.Roslyn.BuildTime"
                      Version="0.6.1"
                      PrivateAssets="all" />`

--------------------------------------------------------------

Then when I try to install: `NuGet\Install-Package dotnet-codegen -Version 0.6.1`

I get an error in the package manager:

```
NuGet\Install-Package : Package 'dotnet-codegen 0.6.1' has a package type 'DotnetCliTool' that is not supported by project 'Models'.
At line:1 char:1
+ NuGet\Install-Package dotnet-codegen -Version 0.6.1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-Package], Exception
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
```

and

` Package 'dotnet-codegen 0.6.1' has a package type 'DotnetCliTool' that is not supported by project 'Models'. `

-------------------------------------------------------------

After searching "Issues" and "Discussions", the more I read about everyone issues, the more I got confused on which packages to import + version, which other tools (if any) to install, project settings/properties to modify etc. So to make everything clear:

1. (sorry here but I'm a NooB) What about the "2.1 sdk" thing? Do I need it to be installed on my computer? Should I reference it and how? Will there be conflicts with my "main sdk", or do I have do downgrade? I use Microsoft.NET.Sdk latest version, .Net7.

2. Once installed, should I ignore some "Warnings" and use languagext-codegen anyways as if it were "okay"?

3. Can anyone show **exactly** what should appear as package reference + version, if I must add other tools and exactly and where to get them, change project settings/prop?

4. Is the documentation still up to date about how to install LgexCodeGen library?

Thank you!

**Comment by StefanBertels** [comment]
```
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFramework>net7.0</TargetFramework>
        <LangVersion>latest</LangVersion>
        <Nullable>enable</Nullable>
    </PropertyGroup>
    <ItemGroup>
        <Using Include="LanguageExt" />
        <Using Include="LanguageExt.Prelude" Static="True" />
        <PackageReference Include="LanguageExt.Core" Version="4.4.0" />
        <PackageReference Include="LanguageExt.CodeGen" Version="4.4.0" />
        <PackageReference Include="CodeGeneration.Roslyn.BuildTime" Version="0.7.5-alpha" PrivateAssets="all" IncludeAssets="runtime; build; native; contentfiles; analyzers" />
        <DotNetCliToolReference Include="dotnet-codegen" Version="0.6.1" />
    </ItemGroup>
</Project>
```

This works for me with .NET Core SDK 2.1 installed, but I'm not sure [whether this is still required](https://github.com/louthy/language-ext/issues/875).
